### PR TITLE
User Model.

### PR DESCRIPTION
**Pull request by leonplondon**
# Summary
Added User Model.
Added Many-To-Many relationship between User & Book.
Added BookAlreadyOwned exception to custom exceptions.

# Trello Card
https://trello.com/c/0q73MoYb/19-crear-modelo-de-user

#Notes
PR pointing to `pr#5-books-crud` because all previous PRs wil be merged soon.

### CODE REVIEW

File: src/main/java/wolox/training/models/User.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Column(nullable = false)
+    private String name;
+
+    @Column(nullable = false)
+    private LocalDate birthDate;
+
+    @OneToOne
+    @JoinColumn(name = "book_id")
+    private Book myBook;
```
Comment (rabinovichm) [inline src/main/java/wolox/training/models/User.java]: Why does an user contains one book separated from the books' list? :thinking:

File: src/main/java/wolox/training/models/Book.java [hunk 2]
```diff
@@ -52,6 +55,9 @@
     @NotEmpty
     private String year;
 
+    @ManyToMany(mappedBy = "books")
+    private List<User> users;
```
Comment (rabinovichm) [inline src/main/java/wolox/training/models/Book.java:61]: Does the book actually needs to contain the users that are holding it? (it's a real question to know what you think)

**Review by rabinovichm — CHANGES_REQUESTED** [review]

File: src/main/java/wolox/training/models/User.java (hunk 1 above)
Comment (leonplondon) [reply src/main/java/wolox/training/models/User.java]: My Bad, was a local test.

File: src/main/java/wolox/training/models/Book.java (hunk 2 above)
Comment (leonplondon) [reply src/main/java/wolox/training/models/Book.java:61]: I think maybe in the future we can use this to get a bidirectional reading between entities.

Comment (leonplondon) [reply src/main/java/wolox/training/models/Book.java:61]: I can remove it, must I?

File: src/main/java/wolox/training/models/User.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    @Column(nullable = false)
+    private String username;
+
+    @Column(nullable = false)
+    private String name;
+
+    @Column(nullable = false)
+    private LocalDate birthDate;
+
+    @ManyToMany(cascade = CascadeType.ALL)
```
Comment (IgnOliveto) [inline src/main/java/wolox/training/models/User.java]: The CascadeType.ALL argument will remove books if you remove users from the data base. More info [here](https://howtodoinjava.com/hibernate/hibernate-jpa-cascade-types/)

File: src/main/java/wolox/training/models/Book.java (hunk 2 above)
Comment (IgnOliveto) [reply src/main/java/wolox/training/models/Book.java:61]: Bidirectional mapping is ok, no need to remove it.

File: src/main/java/wolox/training/models/User.java [hunk 4]
```diff
@@ -32,7 +32,7 @@
     @Column(nullable = false)
     private LocalDate birthDate;
 
-    @ManyToMany(cascade = CascadeType.ALL)
+    @ManyToMany(cascade = {CascadeType.MERGE, CascadeType.PERSIST, CascadeType.DETACH})
```
Comment (IgnOliveto) [inline src/main/java/wolox/training/models/User.java]: In this case the only type of cascade that we need is MERGE, all the other could cause persistence behaviour that we dont want.

File: src/main/java/wolox/training/models/Book.java (hunk 2 above)
Comment (rabinovichm) [reply src/main/java/wolox/training/models/Book.java:61]: Wouldn't it break the [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) by adding a relation that is not needed? I mean, right now it is not necessary to get the users that contain a particular book. @IgnOliveto

Comment (IgnOliveto) [reply src/main/java/wolox/training/models/Book.java:61]: Yes, we  can avoid it but  i dont see much gain avoiding bidirectional mapping, look at this [example](https://www.baeldung.com/hibernate-many-to-many)

**Comment by IgnOliveto** [comment]
Good Job!

**Review by IgnOliveto — APPROVED** [review]

**Review by rabinovichm — APPROVED** [review]

**Pull request merged**